src/ImpactEvaluator.sol

```diff
@@ -134,18 +136,49 @@ contract ImpactEvaluator is AccessControl {
         uint64[] memory scores
     ) private {
         for (uint i = 0; i < addresses.length; i++) {
-            address payable addr = addresses[i];
-            if (addr == 0x000000000000000000000000000000000000dEaD) {
-                // TODO: Shall we emit an event here too?
-                continue;
-            }
-            uint256 amount = (scores[i] * roundReward) / MAX_SCORE;
-            if (addr.send(amount)) {
-                emit Transfer(addr, amount);
-            } else {
-                emit TransferFailed(addr, amount);
-            }
+            balances[addresses[i]] += (scores[i] * roundReward) / MAX_SCORE;
+        }
+    }
+
+    function balanceOf(address account) public view returns (uint) {
+        return balances[account];
+    }
+
+    function _withdraw(
+        address account,
+        address payable target,
+        uint value
+    ) private {
+        require(balances[account] >= value, "Insufficient balance");
+        balances[account] -= value;
+        if (balances[account] == 0) {
+            delete balances[account];
         }
+        require(target.send(value), "Withdrawal failed");
+    }
+
+    function withdraw(address payable target, uint value) public {
+        _withdraw(msg.sender, target, value);
+        emit Withdrawal(msg.sender, target, value);
+    }
+
+    function withdrawOnBehalf(
+        address account,
+        address payable target,
+        uint value,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
```

Comment:
Uh, oh, what is `v`, `r` and `s`? Are these well-known names in EDCSA?

Reply:
Yes, they are well known terms. I didn't bother looking them up, as this was used consistently in all code examples and documentation. See https://github.com/OpenZeppelin/openzeppelin-contracts/blob/94697be8a3f0dfcd95dfb13ffbd39b5973f5c65d/contracts/utils/cryptography/ECDSA.sol